rasa/core/evaluation/stats.py

```diff
@@ -0,0 +1,93 @@
+    """Reads a json marker file.
+
+    Args:
+        path: path to a json file.
+    """
+    path = os.path.abspath(path)
+    with open(path) as json_file:
+        extracted_markers = json.load(json_file)
+        return extracted_markers
+
+
+def compute_summary_stats(data_points: Union[List, np.ndarray]) -> Dict[str, float]:
+    """Computes summary statistics for a given array.
+
+    Computes size, mean, median, min, and max.
+    If size is == 0 returns np.nan for mean, median.
+
+    Args:
+        data_points: can be a numpy array or a list of numbers.
+    """
+    summary_stats = dict()
+    summary_stats["n"] = np.size(data_points)
+
+    if np.size(data_points) > 0:
+        summary_stats["mean"] = np.mean(data_points)
+        summary_stats["median"] = np.median(data_points)
+        summary_stats["min"] = np.min(data_points)
+        summary_stats["max"] = np.max(data_points)
+    else:
+        summary_stats["mean"] = np.nan
+        summary_stats["median"] = np.nan
+        summary_stats["min"] = np.nan
+        summary_stats["max"] = np.nan
+
+    return summary_stats
+
+
+def compute_single_tracker_stats(single_tracker_markers: Dict[str, Any]) -> dict:
+    """Computes summary statistics for a single tracker."""
+    tracker_stats = dict()
+    for marker in single_tracker_markers["markers"]:
+        tracker_stats[marker["marker"]] = compute_summary_stats(
+            marker["num_preceding_user_turns"]
+        )
+    return tracker_stats
+
+
+def compute_multi_tracker_stats(multi_tracker_markers: list) -> Tuple[dict, dict]:
+    """Computes summary statistics for multiple trackers."""
+    overall_stats = dict()
+    overall_stats["num_trackers"] = len(multi_tracker_markers)
+
+    per_tracker_stats = dict()
+    per_marker_values = dict()
+
+    for tracker in multi_tracker_markers:
+        per_tracker_stats[tracker["tracker_ID"]] = compute_single_tracker_stats(tracker)
+        for marker in tracker["markers"]:
+            # append raw values
+            per_marker_values.setdefault(marker["marker"], []).extend(
+                marker["num_preceding_user_turns"]
+            )
+
+    for marker_name in per_marker_values.keys():
+        # compute overall statistics
+        overall_stats[marker_name] = compute_summary_stats(
+            per_marker_values[marker_name]
+        )
+
+    return overall_stats, per_tracker_stats
+
+
+def write_stats(path: Union[Text, Path], stats: dict, per_tracker_stats: dict) -> None:
+    """Outputs statistics to JSON file."""
+    path = os.path.abspath(path)
+    data = {"marker_stats": stats, "tracker_stats": per_tracker_stats}
+    with open(path, "w") as outfile:
+        json_str = json.dumps(data, default=np_encoder, indent=2)
+        outfile.write(json_str)
+
+
+def np_encoder(obj: Any) -> Any:
+    """Encodes numpy array values to make them JSON serializable."""
+    if isinstance(obj, np.generic):
+        return obj.item()
```

Comment:
I'm not sure what convention is but IMO a reference would be good - I think you're right on the types, since this is going to be called on objects that the JSON encoder isn't able to encode to try and handle them.
It's going to implicitly return `None` for anything that gets passed to it that isn't an instance of `np.generic` which in this case is probably fine (since we're we're not passing anything else in that should trip it), but formally it should raise a `TypeError` if it doesn't know what to do.